### PR TITLE
WIP: Implement RV32C instruction decoder

### PR DESCRIPTION
**Pull request by ccs100203**
# RISC-V Compressed Instruction Support

We implemented the compressed instructions for rv32emu-next.
Moreover, We still don't support RV32C.F instructions.
However, it doesn't yet pass the compliance test.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,7 @@ build/DOOM1.WAD*
 build/rv32emu
 *.o
 *.o.d
+
+# Added by Uduru0522
```
Comment (jserv) [inline .gitignore]: You don't have to mention your name since git would track.

File: Makefile [hunk 2]
```diff
@@ -68,4 +69,8 @@ clean:
 distclean: clean
 	$(RM) $(OUT)/DOOM1.WAD $(OUT)/DOOM1.WAD.sha1
 
+debug: clean debug_setting
```
Comment (jserv) [inline Makefile]: This change should only appear in another pull request.

File: io.c [hunk 3]
```diff
@@ -70,7 +70,7 @@ uint32_t memory_read_str(memory_t *m, uint8_t *dst, uint32_t addr, uint32_t max)
 uint32_t memory_read_ifetch(memory_t *m, uint32_t addr)
 {
     const uint32_t addr_lo = addr & mask_lo;
-    assert((addr_lo & 3) == 0);
+    assert((addr_lo & 1) == 0);
```
Comment (jserv) [inline io.c]: Why?

File: riscv.c [hunk 4]
```diff
@@ -770,8 +791,485 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
...
+    rv->PC += rv->inst_len;
+    // enforce zero register
+    if (rd == rv_reg_zero)
+        rv->X[rv_reg_zero] = 0;
+    return true;
+}
+
+static bool c_op_addi4spn(struct riscv_t *rv, uint16_t inst)
+{
+    debug_print("Entered c.addi4spn");
```
Comment (jserv) [inline riscv.c]: You can use GDB script to locate the instruction handlers instead of dumping the craft.

File: riscv.c [hunk 5]
```diff
@@ -770,8 +791,485 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
...
+// CR-type
+static bool c_op_cr(struct riscv_t *rv, uint16_t inst)
+{
+    const uint32_t rs1 = c_dec_rs1(inst);
+    const uint32_t rs2 = c_dec_rs2(inst);
+    const uint32_t rd = rs1;
+
+    switch ((inst & 0x1000) >> 12) {
+    case 0:  
+        if(rs2){
```
Comment (jserv) [inline riscv.c]: Use `clang-format -i` for consistent style.

File: io.c (hunk 3 above)
Comment (ccs100203) [reply io.c]: After executing a compressed instruction, the Program Counter might plus 2, and we can't promise that the last two bits of address always remain 0.  Thus, we modified `(addr_lo & 3)` into `(addr_lo & 1)`.

File: Makefile (hunk 2 above)
Comment (ccs100203) [reply Makefile]: Removed

File: .gitignore [hunk 6]
```diff
@@ -2,3 +2,5 @@ build/DOOM1.WAD*
 build/rv32emu
 *.o
 *.o.d
+test/
```
Comment (jserv) [inline .gitignore]: Rename `test` to directory `tests`.

File: riscv_private.h [hunk 7]
```diff
@@ -210,3 +238,67 @@ static inline uint32_t sign_extend_b(uint32_t x)
...
+static inline uint16_t c_dec_cbtype_imm(uint16_t x){
+    uint16_t temp = 0;
+    //             ....xxxx....xxxx
+    temp |= (x & 0b0000000000011000) >> 2;
+    temp |= (x & 0b0000110000000000) >> 7;
+    temp |= (x & 0b0000000000000100) << 3;
+    temp |= (x & 0b0000000001100000) << 1;
+    temp |= (x & 0b0001000000000000) >> 4;
+    // extend to 16 bit
+    for(int i = 1; i < 8; ++i){
```
Comment (jserv) [inline riscv_private.h]: Use `clang-format -i` to ensure the consistent styie.

File: riscv.c [hunk 8]
```diff
@@ -770,8 +771,440 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
...
+
+    return false;
+}
+
+static bool c_op_bnez(struct riscv_t *rv, uint16_t inst)
+{
+    const uint32_t imm = sign_extend_h(c_dec_cbtype_imm(inst));
+    const uint32_t rs1 = c_dec_rs1c(inst) | 0x08;
+
+    if (rv->X[rs1]) {
```
Comment (jserv) [inline riscv.c]: Rewrite as following:
```c
rv->PC += rv->X[rs1] ? imm : rv->inst_len;
```

File: riscv_private.h (hunk 7 above)
Comment (ccs100203) [reply riscv_private.h]: It's my negligence, I'm sorry.

**Comment by jserv** [comment]
Drop in favor of #11

**Pull request closed without merging**